design-document/design-document.md

```diff
@@ -0,0 +1,131 @@
+
+In a real app, error reporting would be done through an integration (like Sentry), but for the purposes of the challenge `console.error` will suffice. 
+
+There will be a few tests for the key React components and for the key API endpoints, but coverage shouldn't become an obsession.
+
+## Frontend
+
+Let's use React (`create-react-app` for the simplicity of its setup), React Router, and Tailwind on the frontend. Core considerations:
+
+ - **Authentication:** All protected routes will be wrapped in a simple component `<Protected />` which checks whether the user is authenticated. When they're not, the user is taken to the `/sign-in` page, and the original URL is stored in the `state` property on the `location` object of our router. That way, we can redirect the user back to where they intended to go once they've signed in.
+ - **URL as the source of truth:** We will store user's current working directory inside the file tree as URL params, the sorting and filtering criteria as query params (`/browse/my-directory/my-child-directory?sort=name-up&search=holida`).
+ - **Error handling**: Let's pay attention to loading states, errors (at the very least 400, 401, 403, 404, 500) and edge cases such as invalid filtering options, etc.
+ - **Security:** Input and output shall be sanitised (URL params, query params, and the search input field).
+
+## Backend
+
+The server will be a plain `express` app. For the sake of simplicity, let's skip a database and store everything in memory instead.
+
+Since searching and filtering is required on the client only, a handful of API endpoints will suffice:
+
+ - `GET /contents?path={path}` to obtain 1-level deep contents of a directory, returning `{ contents: Item }` where `interface Item { name: string, sizeKb: number | undefined, type: "file" | "dir", items: undefined | Item[] }`
+ - `POST /sign-in` to sign the user in
+ - `POST /sign-out` to sign the user out (`POST` to avoid [prefetching it](https://twitter.com/nick_craver/status/296281730984316928?lang=fa) 🤦) 
+
+## Security
+
+_Note: Whenever generating random, secure tokens, let's use [crypto.randomBytes()](https://nodejs.org/api/crypto.html#cryptorandombytessize-callback)._
+
+Security is, naturally, a huge consideration.
+
+### TLS
+
+To start with, let's use a self-signed certificate to enable `TLS` even on localhost. Unfortunately, this does imply you will need to import the certificate as trusted or "Accept the risk" in your browser. Cheers for your patience in advance 🙏
+
+### Session Management
+
+To avoid depending on external libs, we'll build a simple **session management** ourselves. On signing in, we generate a random `sessionId` for the user. This is used as the `key` in the `sessions` key-value store and also set as a cookie sent back to the client. (See more on security considerations in the [XSS](#cross-site-scripting-xss) and [CSRF](#cross-site-request-forgery-csrf) sections below.)
```

Comment:
I would clarify what would be the cookie attributes (if any).

Reply:
Good point, I did write it in the sections below, but should probably add it to this section, too:

1. `sessionId` gets `httpOnly: true` and `sameSite: lax`
2. `csrfId` gets `sameSite: lax`